# Data Quality food_access_research_atlas

In [1]:
import pandas as pd

In [2]:
#reading in food access data domain 
food_access_2019 = pd.read_csv("food_access_research_atlas.csv")
food_access_2019.head()

CensusTract    State          County  Urban  Pop2010  OHU2010  \
0   1001020100  Alabama  Autauga County      1     1912      693   
1   1001020200  Alabama  Autauga County      1     2170      743   
2   1001020300  Alabama  Autauga County      1     3373     1256   
3   1001020400  Alabama  Autauga County      1     4386     1722   
4   1001020500  Alabama  Autauga County      1    10766     4082   

   GroupQuartersFlag  NUMGQTRS  PCTGQTRS  LILATracts_1And10  ...  \
0                  0       0.0      0.00                  0  ...   
1                  0     181.0      8.34                  1  ...   
2                  0       0.0      0.00                  0  ...   
3                  0       0.0      0.00                  0  ...   
4                  0     181.0      1.68                  0  ...   

   TractSeniors  TractWhite  TractBlack  TractAsian  TractNHOPI  TractAIAN  \
0         221.0      1622.0       217.0        14.0         0.0       14.0   
1         214.0       888.0      1217.0         5.0         0.0        5.0   
2         439.0      2576.0       647.0        17.0         5.0       11.0   
3         904.0      4086.0       193.0        18.0         4.0       11.0   
4        1126.0      8666.0      1437.0       296.0         9.0       48.0   

   TractOMultir  TractHispanic  TractHUNV  TractSNAP  
0          45.0           44.0        6.0      102.0  
1          55.0           75.0       89.0      156.0  
2         117.0           87.0       99.0      172.0  
3          74.0           85.0       21.0       98.0  
4         310.0          355.0      230.0      339.0  

[5 rows x 147 columns]

In [3]:
#Keep just the state of utah data
utah_mask = food_access_2019["State"]  == 'Utah'
food_access_2019 = food_access_2019.loc[utah_mask]
food_access_2019

CensusTract State            County  Urban  Pop2010  OHU2010  \
66425  49001100100  Utah     Beaver County      0     3818     1338   
66426  49001100200  Utah     Beaver County      0     2811      927   
66427  49003960100  Utah  Box Elder County      0     3152      978   
66428  49003960200  Utah  Box Elder County      0     6901     2186   
66429  49003960300  Utah  Box Elder County      1     8186     2558   
...            ...   ...               ...    ...      ...      ...   
67005  49057210900  Utah      Weber County      1     6316     2268   
67006  49057211000  Utah      Weber County      1     3073     1089   
67007  49057211100  Utah      Weber County      1     5785     2093   
67008  49057211201  Utah      Weber County      1     4452     1604   
67009  49057211202  Utah      Weber County      1     6531     2560   

       GroupQuartersFlag  NUMGQTRS  PCTGQTRS  LILATracts_1And10  ...  \
66425                  0      14.0      0.37                  0  ...   
66426                  0      10.0      0.36                  0  ...   
66427                  0       9.0      0.29                  0  ...   
66428                  0       0.0      0.00                  0  ...   
66429                  0      32.0      0.39                  1  ...   
...                  ...       ...       ...                ...  ...   
67005                  0       0.0      0.00                  0  ...   
67006                  0       0.0      0.00                  0  ...   
67007                  0     136.0      2.35                  0  ...   
67008                  0      50.0      1.12                  0  ...   
67009                  0     108.0      1.65                  0  ...   

       TractSeniors  TractWhite  TractBlack  TractAsian  TractNHOPI  \
66425         539.0      3409.0         5.0        54.0        11.0   
66426         293.0      2491.0        11.0        16.0         8.0   
66427         279.0      2981.0         6.0         8.0         4.0   
66428         778.0      6475.0        16.0        75.0        18.0   
66429         717.0      7407.0        29.0       120.0        15.0   
...             ...         ...         ...         ...         ...   
67005        1033.0      5691.0        87.0        79.0        24.0   
67006         429.0      2756.0        67.0        16.0         4.0   
67007         784.0      4918.0       120.0       100.0        27.0   
67008         609.0      4114.0        38.0        68.0         9.0   
67009        1148.0      6001.0        71.0       116.0        26.0   

       TractAIAN  TractOMultir  TractHispanic  TractHUNV  TractSNAP  
66425       43.0         296.0          414.0       41.0      124.0  
66426       33.0         252.0          302.0       10.0       71.0  
66427        8.0         145.0          183.0        5.0       36.0  
66428       22.0         295.0          406.0       32.0      138.0  
66429       41.0         574.0          927.0      150.0      356.0  
...          ...           ...            ...        ...        ...  
67005       31.0         404.0          654.0       44.0      154.0  
67006       24.0         206.0          321.0       27.0      107.0  
67007       42.0         578.0          840.0       75.0      236.0  
67008       28.0         195.0          294.0       92.0       97.0  
67009       21.0         296.0          394.0      200.0      103.0  

[585 rows x 147 columns]

In [5]:
food_access_2019.to_csv('food_access_2019.csv', index=False)

In [12]:
#keeping the columns I want
#Field	LongName
#CensusTract	Census tract
#State	State
#County	County
#Urban	Urban tract
#POP2010	Population, tract total
#PovertyRate	Tract poverty rate
#MedianFamilyIncome	Tract median family income
#LA1and10	Low access tract at 1 mile for urban areas and 10 miles for rural areas
#LAhalfand10	Low access tract at 1/2 mile for urban areas and 10 miles for rural areas
#LA1and20	Low access tract at 1 mile for urban areas and 20 miles for rural areas
#LATracts_half	Low access tract at 1/2 mile
#LATracts1	Low access tract at 1 mile
#LATracts10	Low access tract at 10 miles
#LATracts20	Low access tract at 20 miles
#TractWhite	Tract White population, number
#TractBlack	Tract Black or African American population, number
#TractAsian	Tract Asian population, number
#TractNHOPI	Tract Native Hawaiian and Other Pacific Islander population, number
#TractAIAN	Tract American Indian and Alaska Native population, number
#TractOMultir	Tract Other/Multiple race population, number
#TractHispanic	Tract Hispanic or Latino population, number

food_access_2019 = food_access_2019[['CensusTract', 'State', 'County', 'Urban', 'Pop2010', 'PovertyRate',
                                'MedianFamilyIncome', 'LA1and10', 'LAhalfand10', 'LA1and20', 
                                'LATracts_half', 'LATracts1', 'LATracts10', 'LATracts20',
                                'TractWhite', 'TractBlack', 'TractAsian', 
                                'TractNHOPI', 'TractAIAN', 'TractOMultir', 'TractHispanic']]
food_access_2019.head()

CensusTract State            County  Urban  Pop2010  PovertyRate  \
66425  49001100100  Utah     Beaver County      0     3818          8.7   
66426  49001100200  Utah     Beaver County      0     2811          7.8   
66427  49003960100  Utah  Box Elder County      0     3152          5.6   
66428  49003960200  Utah  Box Elder County      0     6901          8.3   
66429  49003960300  Utah  Box Elder County      1     8186         16.1   

       MedianFamilyIncome  LA1and10  LAhalfand10  LA1and20  ...  LATracts1  \
66425             65332.0         0            0         0  ...          0   
66426             68112.0         1            1         0  ...          0   
66427             66284.0         1            1         1  ...          0   
66428             72953.0         0            0         0  ...          0   
66429             53281.0         1            1         1  ...          1   

       LATracts10  LATracts20  TractWhite  TractBlack  TractAsian  TractNHOPI  \
66425           0           0      3409.0         5.0        54.0        11.0   
66426           1           0      2491.0        11.0        16.0         8.0   
66427           1           1      2981.0         6.0         8.0         4.0   
66428           0           0      6475.0        16.0        75.0        18.0   
66429           0           0      7407.0        29.0       120.0        15.0   

       TractAIAN  TractOMultir  TractHispanic  
66425       43.0         296.0          414.0  
66426       33.0         252.0          302.0  
66427        8.0         145.0          183.0  
66428       22.0         295.0          406.0  
66429       41.0         574.0          927.0  

[5 rows x 21 columns]

# Attribute domain constraints

Basic aggregate statistics for the food_access_2019 data domain 

The descriptive statistics seem reasonable. The Low access columns and urban all have a min of 0 and Max of 1 as expected since they are binary results. All other columns seem reasonable. 

In [6]:
food_access_2019.describe()

CensusTract       Urban       Pop2010      OHU2010  GroupQuartersFlag  \
count  5.850000e+02  585.000000    585.000000   585.000000         585.000000   
mean   4.903625e+10    0.851282   4724.589744  1500.328205           0.005128   
std    1.537811e+07    0.356115   2051.516805   633.025840           0.071489   
min    4.900110e+10    0.000000    103.000000     0.000000           0.000000   
25%    4.903510e+10    1.000000   3462.000000  1081.000000           0.000000   
50%    4.903511e+10    1.000000   4452.000000  1440.000000           0.000000   
75%    4.904900e+10    1.000000   5788.000000  1821.000000           0.000000   
max    4.905721e+10    1.000000  21591.000000  5699.000000           1.000000   

          NUMGQTRS    PCTGQTRS  LILATracts_1And10  LILATracts_halfAnd10  \
count   585.000000  585.000000         585.000000            585.000000   
mean     78.892308    1.969675           0.083761              0.263248   
std     311.228873    8.171626           0.277266              0.440773   
min       0.000000    0.000000           0.000000              0.000000   
25%       0.000000    0.000000           0.000000              0.000000   
50%       0.000000    0.000000           0.000000              0.000000   
75%      43.000000    0.990000           0.000000              1.000000   
max    3840.000000  100.000000           1.000000              1.000000   

       LILATracts_1And20  ...  TractSeniors    TractWhite  TractBlack  \
count         585.000000  ...    585.000000    585.000000  585.000000   
mean            0.076923  ...    426.430769   4067.623932   50.063248   
std             0.266697  ...    283.533790   1883.151067   56.359546   
min             0.000000  ...      1.000000     43.000000    0.000000   
25%             0.000000  ...    229.000000   2877.000000   16.000000   
50%             0.000000  ...    385.000000   3782.000000   33.000000   
75%             0.000000  ...    554.000000   4942.000000   64.000000   
max             1.000000  ...   2240.000000  20051.000000  497.000000   

        TractAsian  TractNHOPI    TractAIAN  TractOMultir  TractHispanic  \
count   585.000000  585.000000   585.000000    585.000000     585.000000   
mean     94.504274   41.972650    56.285470    414.140171     612.547009   
std     100.783032   61.731544   208.109533    397.775463     606.079877   
min       0.000000    0.000000     0.000000      0.000000       1.000000   
25%      30.000000    9.000000    15.000000    168.000000     232.000000   
50%      68.000000   21.000000    29.000000    272.000000     385.000000   
75%     123.000000   47.000000    52.000000    511.000000     785.000000   
max    1064.000000  563.000000  3771.000000   2467.000000    3496.000000   

        TractHUNV   TractSNAP  
count  585.000000  585.000000  
mean    67.500855  117.986325  
std     81.596089  109.475242  
min      0.000000    0.000000  
25%     16.000000   40.000000  
50%     40.000000   86.000000  
75%     91.000000  164.000000  
max    656.000000  691.000000  

[8 rows x 145 columns]

In [7]:
import matplotlib
import matplotlib.pyplot as plt
plt.style.use('ggplot')
# This next line tells jupyter to render the images inline
%matplotlib inline

The value counts for urban show that there are more urban census tracts than rural in the state of utah. LAhalfand10 is the varaible for low access tract at 1/2 mile for urban areas and 10 miles for rural areas. There are more 1's I assume mean yes than 0's which I assume mean no. There are 585 census tracts. Each row in the dataset represents a unique census tract. Salt Lake city county has the most census tracts with 211. 

In [8]:
food_access_2019["Urban"].value_counts()

Urban
1    498
0     87
Name: count, dtype: int64

In [9]:
food_access_2019["LAhalfand10"].value_counts()

LAhalfand10
1    496
0     89
Name: count, dtype: int64

In [10]:
food_access_2019['CensusTract'].value_counts()

CensusTract
49001100100    1
49049000707    1
49049000203    1
49049000204    1
49049000205    1
              ..
49035111304    1
49035111305    1
49035111306    1
49035111400    1
49057211202    1
Name: count, Length: 585, dtype: int64

In [11]:
food_access_2019.groupby(['County']).describe()

CensusTract                                             \
                        count          mean            std           min   
County                                                                     
Beaver County             2.0  4.900110e+10      70.710678  4.900110e+10   
Box Elder County         11.0  4.900396e+10     240.757970  4.900396e+10   
Cache County             26.0  4.900504e+10  192064.627410  4.900500e+10   
Carbon County             5.0  4.900700e+10     207.364414  4.900700e+10   
Daggett County            1.0  4.900996e+10            NaN  4.900996e+10   
Davis County             54.0  4.901113e+10     568.639313  4.901113e+10   
Duchesne County           3.0  4.901394e+10     152.752523  4.901394e+10   
Emery County              3.0  4.901598e+10     152.752523  4.901598e+10   
Garfield County           2.0  4.901700e+10      70.710678  4.901700e+10   
Grand County              2.0  4.901900e+10      70.710678  4.901900e+10   
Iron County               8.0  4.902111e+10     226.882246  4.902111e+10   
Juab County               2.0  4.902301e+10      70.710678  4.902301e+10   
Kane County               2.0  4.902513e+10      70.710678  4.902513e+10   
Millard County            3.0  4.902797e+10     100.000000  4.902797e+10   
Morgan County             2.0  4.902997e+10      70.710678  4.902997e+10   
Piute County              1.0  4.903196e+10            NaN  4.903196e+10   
Rich County               1.0  4.903395e+10            NaN  4.903395e+10   
Salt Lake County        211.0  4.903511e+10    4454.360347  4.903510e+10   
San Juan County           4.0  4.903796e+10   20842.424683  4.903794e+10   
Sanpete County            5.0  4.903997e+10     158.113883  4.903997e+10   
Sevier County             5.0  4.904198e+10     158.113883  4.904198e+10   
Summit County            13.0  4.904396e+10      96.823723  4.904396e+10   
Tooele County            10.0  4.904513e+10     199.930794  4.904513e+10   
Uintah County             6.0  4.904796e+10   11480.270637  4.904794e+10   
Utah County             127.0  4.904900e+10    4193.600534  4.904900e+10   
Wasatch County            4.0  4.905196e+10    9867.454923  4.905194e+10   
Washington County        21.0  4.905327e+10     520.621501  4.905327e+10   
Wayne County              1.0  4.905598e+10            NaN  4.905598e+10   
Weber County             50.0  4.905721e+10    4824.614888  4.905720e+10   

                                                                           \
                            25%           50%           75%           max   
County                                                                      
Beaver County      4.900110e+10  4.900110e+10  4.900110e+10  4.900110e+10   
Box Elder County   4.900396e+10  4.900396e+10  4.900396e+10  4.900396e+10   
Cache County       4.900500e+10  4.900500e+10  4.900500e+10  4.900598e+10   
Carbon County      4.900700e+10  4.900700e+10  4.900700e+10  4.900700e+10   
Daggett County     4.900996e+10  4.900996e+10  4.900996e+10  4.900996e+10   
Davis County       4.901113e+10  4.901113e+10  4.901113e+10  4.901113e+10   
Duchesne County    4.901394e+10  4.901394e+10  4.901394e+10  4.901394e+10   
Emery County       4.901598e+10  4.901598e+10  4.901598e+10  4.901598e+10   
Garfield County    4.901700e+10  4.901700e+10  4.901700e+10  4.901700e+10   
Grand County       4.901900e+10  4.901900e+10  4.901900e+10  4.901900e+10   
Iron County        4.902111e+10  4.902111e+10  4.902111e+10  4.902111e+10   
Juab County        4.902301e+10  4.902301e+10  4.902301e+10  4.902301e+10   
Kane County        4.902513e+10  4.902513e+10  4.902513e+10  4.902513e+10   
Millard County     4.902797e+10  4.902797e+10  4.902797e+10  4.902797e+10   
Morgan County      4.902997e+10  4.902997e+10  4.902997e+10  4.902997e+10   
Piute County       4.903196e+10  4.903196e+10  4.903196e+10  4.903196e+10   
Rich County        4.903395e+10  4.903395e+10  4.903395e+10  4.903395e+10   
Salt Lake County   4.903511e+10  4.903511e+10

In [17]:
food_access_2019["County"].value_counts()

County
Salt Lake County     211
Utah County          127
Davis County          54
Weber County          50
Cache County          26
Washington County     21
Summit County         13
Box Elder County      11
Tooele County         10
Iron County            8
Uintah County          6
Carbon County          5
Sanpete County         5
Sevier County          5
Wasatch County         4
San Juan County        4
Duchesne County        3
Emery County           3
Millard County         3
Beaver County          2
Kane County            2
Juab County            2
Grand County           2
Garfield County        2
Morgan County          2
Rich County            1
Piute County           1
Daggett County         1
Wayne County           1
Name: count, dtype: int64

Optionality was described as number of missing values in the paper. There are two missing values in the MedianFamilyIncome column. They both are in salt lake county. 


In [20]:
#checking for null values
food_access_2019.isnull().sum()

CensusTract           0
State                 0
County                0
Urban                 0
Pop2010               0
PovertyRate           0
MedianFamilyIncome    2
LA1and10              0
LAhalfand10           0
LA1and20              0
LATracts_half         0
LATracts1             0
LATracts10            0
LATracts20            0
TractWhite            0
TractBlack            0
TractAsian            0
TractNHOPI            0
TractAIAN             0
TractOMultir          0
TractHispanic         0
dtype: int64

In [28]:
#Finding the rows for the MedianFamilyIncome that are missing

food_access_2019.loc[food_access_2019['MedianFamilyIncome'].isnull()]

CensusTract State            County  Urban  Pop2010  PovertyRate  \
66672  49035112818  Utah  Salt Lake County      1     3840          0.0   
66782  49043964203  Utah     Summit County      0      103          0.0   

       MedianFamilyIncome  LA1and10  LAhalfand10  LA1and20  ...  LATracts1  \
66672                 NaN         1            1         1  ...          1   
66782                 NaN         1            1         1  ...          0   

       LATracts10  LATracts20  TractWhite  TractBlack  TractAsian  TractNHOPI  \
66672           0           0      2870.0       310.0        13.0        20.0   
66782           1           1        89.0         0.0         0.0         1.0   

       TractAIAN  TractOMultir  TractHispanic  
66672      165.0         462.0          816.0  
66782        0.0          13.0           12.0  

[2 rows x 21 columns]

# Relational integrity rules

The paper refers to checking for identity. In this data set that means making sure the CensusTract are unique. Using food_access_2019["CensusTract"].describe() shows that there are 585 rows in the table and 585 unique census tracts. 

In [33]:
food_access_2019["CensusTract"].describe()

count    5.850000e+02
mean     4.903625e+10
std      1.537811e+07
min      4.900110e+10
25%      4.903510e+10
50%      4.903511e+10
75%      4.904900e+10
max      4.905721e+10
Name: CensusTract, dtype: float64

The paper refers to checking for reference. Will need to check that the census tract can be used to join the individual datasets together. 

# Historical data rules 

Historical data rules according to the paper are rules that look at "time-varying data". This dataset does not include dates to do specfic date checks with. According to the source overview the data is based on the 2019 list of supermarkets. https://www.ers.usda.gov/data-products/food-access-research-atlas/about-the-atlas/

# State dependent objects rules

The values for urban and for the low access columns can only be 0 or 1 not both or a value in between. Looking at the values counts in the attribute doman constraint section shows that this is the the case. The value counts show that 0 and 1 are the only values in the column and the totals add up to 585 as expected. 

# Attribute dependency rules

To assess attribute dependency rules we should test that redundant attributes in different data sources have identical values. 

The sum of the races equals the the total population for the census tract. 

In [48]:
#Derived attributes: aggregated valu must equal the total of the atmoic level values.
#Checking if the sum of the race categories equals the total population
food_access_2019['total_race_count'] = food_access_2019['TractWhite'] + food_access_2019['TractBlack'] + food_access_2019['TractAsian'] +  food_access_2019['TractNHOPI'] + food_access_2019['TractAIAN'] + food_access_2019['TractOMultir']

food_access_2019[['Pop2010','total_race_count']]

Pop2010  total_race_count
66425     3818            3818.0
66426     2811            2811.0
66427     3152            3152.0
66428     6901            6901.0
66429     8186            8186.0
...        ...               ...
67005     6316            6316.0
67006     3073            3073.0
67007     5785            5785.0
67008     4452            4452.0
67009     6531            6531.0

[585 rows x 2 columns]

# Conclusion

Overall, the quality of data is good in the the food access 2019 data set. There are two  missing values in the medianfamilyincome. The outputs in the columns are as expected for example they are a 0 or a 1 when they should be and the tsum of the aggregrate race data matches the total population data. More assessment should be done when we combine the datasets to make sure the values are the same across sources for redundant columns. 

The facets dataset used the Low access to healthy food (1/2 mile). Since Utah has both rural and urban census tracts I think we should use the Low access tract at 1/2 mile for urban areas and 10 miles for rural areas column. That is the column I focused on in this assessment. There are other options to explore. 



# Resources

Description of the columns: https://www.ers.usda.gov/data-products/food-access-research-atlas/documentation/#distance

Download the data: https://www.ers.usda.gov/data-products/food-access-research-atlas/download-the-data/ in the zip files there is one spreadsheet that gives a brief explanation of the columns and another excel file with the actual data

See the right hand side of both the links above for additional information about this data source. 
